### PR TITLE
[Transform] [Quantization] Add transforms to compressed tensors

### PR DESCRIPTION
**Pull request by kylesayrs**
# Online Hadamard Rotations #

Screenshot 2025-08-26 at 20 22 12

## Purpose ##
* Support online linear transforms for models compressed with CompressedTensors

### Added Transforms Weight Loading ###
* Implemented `SharedWeightParameter`
  * This class implements a registry where tensors can be shared. Loading with shared tensors is required for models with many transforms such as QuIP
* Before weight loading, a `HadamardTransform` module is attached to linear layers to load transform weights which utilizes `SharedWeightParameter` in order to load weight partitions as separate tensors

### Added Transforms Apply ###
* Add `CompressedTensorsLinearTransformMethod` which wraps `CompressedTensorsLinearMethod` and `UnquantizedLinearMethod` adds input and output transforms to either side of the original `apply` method
  * Unfortunately, because https://github.com/vllm-project/vllm/pull/23036 has not landed, we must use a [hack](https://github.com/vllm-project/vllm/pull/22486/files#diff-83f6e833e3da91b336bf076817f7543c83c5bc41f398824772ba0f101c2c3481R56-R75) to switch back to weight_loader_v1 if the method being wrapped is the UnquantizedLinearMethod
  * This class has template functionality for selecting which kernel to use when applying transforms

### Misc Changes ###
* Add transform config parsing to CompressedTensors
* Define `_shard_id_as_int` on `BasevLLMParameter`. This is so that its implementation can be shared with `SharedWeightParameter`
* Add test utility `calculate_prompt_perplexity` for checking model coherence
* Added `weight_loader_v1` to `QKVCrossParallelLinear` to support hack described above (and https://github.com/vllm-project/vllm/pull/23036)

## Testing ##
* Added `test_compressed_tensors_transforms_perplexity` tests for SpinQuantR1R2R4 and QuIP
* Locally tested unquantized QuIP and SpinQuantR1R2R4 models

### CODE REVIEW

**Review by yewentao256 — COMMENTED** [review]
Let's run CI and see if it is correct

File: vllm/model_executor/layers/quantization/compressed_tensors/transform/linear.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+from vllm.model_executor.layers.linear import (WEIGHT_LOADER_V2_SUPPORTED,
+                                               LinearMethodBase,
+                                               QKVCrossParallelLinear)
+from vllm.model_executor.layers.quantization.compressed_tensors.transform.module import (  # noqa: E501
+    HadamardTransform)
+from vllm.model_executor.layers.quantization.compressed_tensors.transform.utils import (  # noqa: E501
+    TransformTuple)
+
+
+class CompressedTensorsLinearTransformMethod(LinearMethodBase):
```
Comment (dsikka) [inline vllm/model_executor/layers/quantization/compressed_tensors/transform/linear.py:21]: Maybe a docstring describing what this class is doing?

File: vllm/model_executor/layers/quantization/compressed_tensors/transform/linear.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+            for tfm in self.output_tfms.values():
+                if tfm.scheme_name != scheme_name:
+                    raise ValueError("Must have same scheme name")
+                if tfm.args.location != location:
+                    raise ValueError("Must have same location")
+
+        return self.input_tfms, self.output_tfms
+
+
+def get_linear_transform_schemes(
```
Comment (dsikka) [inline vllm/model_executor/layers/quantization/compressed_tensors/transform/linear.py:162]: docstring

File: vllm/model_executor/layers/quantization/compressed_tensors/transform/linear.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+        # get weight loader for transforms
+        weight_loader: Callable = extra_weight_attrs.get(
+            "weight_loader")  # type: ignore[assignment]
+
+        # HACK: UnquantizedLinearMethod does not support weight loader v2, but
+        # transforms (specifically PartitionedLinearWeightParameter) requires
+        # weight loader v2. Until UnquantizedLinearMethod supports v2, we must
+        # hack around this by getting weight loader v1 so ULM can load correctly
+        quant_method_name = self.quant_method.__class__.__name__
```
Comment (dsikka) [inline vllm/model_executor/layers/quantization/compressed_tensors/transform/linear.py:64]: So if I understand correctly, this is to load transforms that target  unquantized layers, such as the embedding layer?

File: vllm/model_executor/layers/linear.py [hunk 4]
```diff
@@ -1469,7 +1471,7 @@ def __init__(self,
             self.bias = torch.nn.Parameter()
             set_weight_attrs(self.bias, {
                 "output_dim": 0,
-                "weight_loader": self.weight_loader,
+                "weight_loader": self.weight_loader_v1,
```
Comment (dsikka) [inline vllm/model_executor/layers/linear.py:1475]: confusing naming given the non-v2 loader is v1

**Review by yewentao256 — COMMENTED** [review]
Thanks for the work! Could you also report `lm_eval` for E2E accuracy and `vllm bench` for E2E throughput metrics? (comparison with main)

File: vllm/model_executor/layers/quantization/compressed_tensors/transform/linear.py (hunk 3 above)
Comment (kylesayrs) [reply vllm/model_executor/layers/quantization/compressed_tensors/transform/linear.py:64]: We don't support online rotations for embedding layers as of now. But you are correct, this code catches the case where a user runs an unquantized, transformed model (for testing and validation)

**Comment by dsikka** [comment]
> Thanks for the work! Could you also report `lm_eval` for E2E accuracy and `vllm bench` for E2E throughput metrics? (comparison with main)

The small lm-eval model test run automatically through the CI.
We can enable the large model tests additionally. That should be sufficient for lm-eval runs.

File: vllm/model_executor/layers/linear.py (hunk 4 above)
Comment (kylesayrs) [reply vllm/model_executor/layers/linear.py:1475]: > confusing naming given the non-v2 loader is v1

I'm confused what the confusion is here. On `QKVCrossParallelLinear`, the default weight loader chooses whether to use v2 or v1. Other linear module implementations implicitly force the bias parameter to be loaded with v1. This module is missing a v1 implementation.

Given that the name `weight_loader` is already being used to describe a loader that uses either v1 or v2, I think `weight_loader_v1` is an appropriate name to describe a loader that exclusively uses v1.

**Comment by kylesayrs** [comment]
It seems that adding [these tests](https://github.com/vllm-project/vllm/pull/22486/files#diff-b1bb162888fda73aa6256e3b7a6794d23f881b91994a151f908d77df5dca4b10R689-R712) cause subsequent tests to go OOM.

@mgoin @yewentao256 Do you have any suggestions as to how to mitigate this?

EDIT: this was an issue with tensors in the shared tensors registry not being cleaned up between runs

File: vllm/model_executor/layers/linear.py (hunk 4 above)
Comment (dsikka) [reply vllm/model_executor/layers/linear.py:1475]: The other linear base methods have methods called  [weight_laoder](https://github.com/neuralmagic/vllm/blob/09b4c1bdf58dcc5c0aa2ffa8451f3733644ea0c1/vllm/model_executor/layers/linear.py#L1304)` (which is the v1 weight loader) or [weight_loader_v2](https://github.com/neuralmagic/vllm/blob/09b4c1bdf58dcc5c0aa2ffa8451f3733644ea0c1/vllm/model_executor/layers/linear.py#L1341). 

You are now adding a third method called weight_loader_v1, which is confusing given the old weight_loader is v1.

If you want to keep it, I would add something to make it clear that this isn't some third weight loader.

Comment (kylesayrs) [reply vllm/model_executor/layers/linear.py:1475]: The problem here is that `weight_loader` is misnamed and should be renamed to `weight_loader_v2`. Given the short timeline for this PR, I'd like to address this in a follow up rather than risk issues with name changes for now

I think the naming and code comments makes it clear what this function is doing.

File: tests/conftest.py [hunk 5]
```diff
@@ -991,11 +993,34 @@ def generate_greedy_logprobs(
...
+    def generate_prompt_perplexity(self, prompts: list[str]) -> list[float]:
+        outputs = self.generate_greedy_logprobs(prompts,
+                                                max_tokens=1,
+                                                num_logprobs=None,
+                                                num_prompt_logprobs=0)
+
+        perplexities = []
+        for output in outputs:
+            output = cast(TokensTextLogprobsPromptLogprobs, output)
+            token_datas = cast(list[Optional[dict[int, Logprob]]], output[3])
```
Comment (mgoin) [inline tests/conftest.py:1011]: Why do we need this casting? Could we get the same result by asserting isinstance?

File: tests/conftest.py [hunk 6]
```diff
@@ -991,11 +993,34 @@ def generate_greedy_logprobs(
                                         videos=videos,
                                         **kwargs)
 
+    def generate_prompt_perplexity(self, prompts: list[str]) -> list[float]:
```
Comment (mgoin) [inline tests/conftest.py:996]: Would be nice to have a docstring to detail the operation here

File: tests/conftest.py [hunk 7]
```diff
@@ -991,11 +993,34 @@ def generate_greedy_logprobs(
...
+        outputs = self.generate_greedy_logprobs(prompts,
+                                                max_tokens=1,
+                                                num_logprobs=None,
```
Comment (mgoin) [inline tests/conftest.py:1005]: What does None mean here?

File: vllm/model_executor/parameter.py [hunk 8]
```diff
@@ -373,6 +378,141 @@ class BlockQuantScaleParameter(_ColumnvLLMParameter, RowvLLMParameter):
     pass
 
 
+class SharedWeightParameter(BasevLLMParameter):
+    """
+    Parameter for weights with many shared tensors across a model
+
+    For example, when applying transforms to the "gate" and "up" partitions of
+    `MergedColumnParallelLinear`, the tranform weights must stay separate
```
Comment (mgoin) [inline vllm/model_executor/parameter.py]:
```suggestion
    `MergedColumnParallelLinear`, the transform weights must stay separate
```

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py [hunk 9]
```diff
@@ -103,18 +113,27 @@ def get_quant_method(
     ) -> Optional["QuantizeMethodBase"]:
         from vllm.attention.layer import Attention  # Avoid circular import
 
-        # Check if the layer is skipped for quantization.
-        # TODO (@robertgshaw2): support module names
-        if should_ignore_layer(prefix,
-                               ignore=self.ignore,
-                               fused_mapping=self.packed_modules_mapping):
-            return UnquantizedLinearMethod()
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:111]: I don't remember the significance of having this function return None vs UnquantizedLinearMethod - why is it okay to remove the ignoring here?

File: tests/conftest.py (hunk 5 above)
Comment (kylesayrs) [reply tests/conftest.py:1011]: `isinstance` doesn't work for composite types (for example tuples). I might be speaking imprecisely here, but TLDR I tried with these types and it doesn't work.

File: tests/conftest.py (hunk 6 above)
Comment (kylesayrs) [reply tests/conftest.py:996]: I didn't add one since the others don't have any, but I agree that this would be useful

File: tests/conftest.py (hunk 7 above)
Comment (kylesayrs) [reply tests/conftest.py:1005]: `num_logprobs=None` means do not produce logprobs for decoded tokens, whereas `num_prompt_logprobs=0` means produce one logprob for each prefill (prompt) token

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py (hunk 9 above)
Comment (kylesayrs) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:111]: `get_quant_method` only returns None if the module is not a linear, attention, or moe. If it's a linear, it will always return either UnquantizedLinearMethod or CompressedTensorsLinearMethod (theoretically).

This ignoring has been moved to within the `get_scheme` method so that quantization ignores won't affect transform matching.

Comment (kylesayrs) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:111]: FYI this will also be replaced with a CT util later
https://github.com/vllm-project/vllm/pull/23597

**Review by mgoin — APPROVED** [review]
